Document `/http` (#550)
libp2p/specs: http/transport-component.md

## What is an HTTP transport

An HTTP transport is simply a node that can speak some standardized version of
HTTP. Intuitively if you can `curl` it with HTTP, then it speaks HTTP.

Most environments will have a way to create an HTTP Client and Server, and the
specific HTTP version used will be opaque. We use the `/http` component at the
end of the multidadr to signal that this server supports an HTTP transport. The
end user agent decides on HTTP version to use, based on the multiaddr prefix,
application, server negotiation, and specific use case. This follows what
existing `http://` URL implementations do.

## Multiaddr representation

The multiaddr of a node with an HTTP transport ends with `/http` and is prefixed
by information that would let an HTTP client know how to reach the server
(remember that multiaddrs are [interpreted right to
left](https://github.com/multiformats/multiaddr#interpreting-multiaddrs)). 

The following are examples of multiaddrs for HTTP transport capable nodes:

* `/dns/example.com/tls/http`
* `/ip4/1.2.3.4/tcp/443/tls/http`
* `/ip6/2001:0db8:85a3:0000:0000:8a2e:0370:7334/tcp/443/tls/http`
* `/ip4/1.2.3.4/udp/50781/tls/http`

> ## HTTP Paths

Author: I almost want to remove this section since it's kind of a tangent. It's useful in making the case why `/httppath` shouldn't be appended to `/http` and what you can do instead, but it also muddies the spec.

Reviewer: Hm.. maybe replace the tangent with a list of HTTP Semantics that one would be tempted to expose after `/http`, like I've hinted in https://github.com/ipni/go-libipni/issues/42#issuecomment-1570503551?

> After all, paths are not special, it is only one of many attributes that may impact HTTP endpoint's behavior.  If we allow HTTP paths, the unanswered question is: how one adds basic auth or bearer token headers? or a cookie? or an Accept header for content type version negotiation? [Do we add `/foo` for every HTTP feature? Where is the line in the sand?]
> 
> There are prior discussions in
> - https://github.com/multiformats/multiaddr/issues/63
> - https://github.com/multiformats/multiaddr/issues/87

Reviewer 2: Should this section make recommendations on how paths should be handled? The HTTP rest APIs commonly support this. I think at the very least there should be recommendations on how should one handle HTTP API path prefixes when expressed as multiaddr.

Reviewer 2: Many thanks for adding this @MarcoPolo 

Using private ranges to specify path is certainly one way to do it. However, I am conscious of the increase in barrier for adoption. This is something that hopefully we can iterate on with the help of the lip2p community as HTTP over libp2p matures.

transport. That will be defined in a future document.

This exists to clarify the role of the `/http` component in Multiaddrs early to
avoid confusion and conflicting interpretations.

## What is an HTTP transport

An HTTP transport is simply a node that can speak some standardized version of
HTTP. Intuitively if you can `curl` it with HTTP, then it speaks HTTP.

Most environments will have a way to create an HTTP Client and Server, and the
specific HTTP version used will be opaque. We use the `/http` component at the
end of the multidadr to signal that this server supports an HTTP transport. The
end user agent decides on HTTP version to use, based on the multiaddr prefix,
application, server negotiation, and specific use case. This follows what
existing `http://` URL implementations do.

## Multiaddr representation

The multiaddr of a node with an HTTP transport ends with `/http` and is prefixed
by information that would let an HTTP client know how to reach the server
(remember that multiaddrs are [interpreted right to
left](https://github.com/multiformats/multiaddr#interpreting-multiaddrs)). 

The following are examples of multiaddrs for HTTP transport capable nodes:

> * `/dns/example.com/tls/http`

Reviewer 3: ```suggestion
* `/dns/example.com/tcp/443/tls/http`
```

Author: Let's keep the discussion on https://github.com/libp2p/specs/pull/550#discussion_r1213520210

* `/ip4/1.2.3.4/tcp/443/tls/http`
* `/ip6/2001:0db8:85a3:0000:0000:8a2e:0370:7334/tcp/443/tls/http`

> * `/ip4/1.2.3.4/udp/50781/tls/http`

Reviewer 3: ```suggestion
* `/ip4/1.2.3.4/udp/50781/quic/http`
```

# HTTP Transport Component <!-- omit in toc -->

| Lifecycle Stage | Maturity      | Status | Latest Revision |
| --------------- | ------------- | ------ | --------------- |
| 1A              | Working Draft | Active | r0, 2023-05-31  |

Authors: [@marcopolo]

Interest Group: [@marcopolo], [@mxinden], [@marten-seemann]

[@marcopolo]: https://github.com/marcopolo
[@mxinden]: https://github.com/mxinden
[@marten-seemann]: https://github.com/marten-seemann

## Table of Contents <!-- omit in toc -->
- [Context](#context)
- [What is an HTTP transport](#what-is-an-http-transport)
- [Multiaddr representation](#multiaddr-representation)
- [HTTP Paths](#http-paths)

## Context

This document is only about advertising support for an HTTP transport. It
doesn't make any assertions about how libp2p should interact with that
transport. That will be defined in a future document.

This exists to clarify the role of the `/http` component in Multiaddrs early to
avoid confusion and conflicting interpretations.

## What is an HTTP transport

An HTTP transport is simply a node that can speak some standardized version of

> HTTP that is at least HTTP/1.1. Currently that means HTTP/1.1, HTTP/2, and

Reviewer 3: Is it necessary to specify this? HTTP/1.0 is not really in use any more, but it would be nice if we didn't have to say anything here.

Author: Yeah, I can remove the list. We should still specify at least 1.1.

Reviewer 3: Why?

Author: Because HTTP/1.0 is still in use (It's the default for nginx: https://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_http_version). We should be explicit that we only support at least 1.1

Reviewer 3: That seems like an argument for not specifying it. Let the HTTP layer do the HTTP things.

Author: Do you want to support HTTP/1.0? I don't. If we don't set a minimum requirement we are effectively saying we support HTTP/1.0

Reviewer 3: We’re not. We’re just saying we don’t get involved in HTTP internals. How you configure your HTTP client and server is on you.

Author: * We're defining a standard here that folks will reference when they want their nodes to interoperate with the network.
* We don't want them to use HTTP/1.0.

What parts do you disagree with?

Reviewer 3: We’re also not telling people which cipher suites they need to support when using TLS / QUIC. Somebody could configure some obscure cipher suites, which would make cipher suite negotiation fail during the handshake.

Author: I don't feel strongly on this. I will remove the HTTP/1.1 requirement.

HTTP/3. Intuitively if you can `curl` it with HTTP, then it speaks HTTP.

Most environments will have a way to create an HTTP Client and Server, and the
specific HTTP version used will be opaque. The client will negotiate (or [learn

> about](https://www.rfc-editor.org/rfc/rfc9114.html#section-3.1.1)) the specific HTTP version to use when communicating

Reviewer 3: This is probably not the best document to reference any more. Nowadays you'd look at the SVCB DNS record, and soon the AltSvcB HTTP header.

Author: Could you link a better reference? I found https://www.ietf.org/archive/id/draft-ietf-dnsop-svcb-https-12.html but it's still a draft.

Author: Not necessary. Just helpful. we can remove that link.

with the HTTP server.

For this opaque case, we use the `/http` component at the end of the multidadr.
If a node wants to advertise specific HTTP versions it may advertise a

> `/http-1.1`, `/http-2`, or `/h3` instead. Although these components are not

Reviewer 3: Why would you do that? HTTP has its own version negotiation mechanism (ALPN for HTTP/1.1 and HTTP/2, more complicated (see above) for HTTP/3. I don't think we need to do anything here.

Author: You can avoid paying an extra round trip for version negotiation. If your server prefers HTTP/2 and always upgrades clients to that, you can tell clients ahead of time to skip the round trip cost for the upgrade to HTTP/2 and just start with http/2.

Reviewer 3: There’s no round trip cost. ALPN negotiation happens during the TLS handshake.

If you made the argument that SVCB contains a protocol hint, that would make me reconsider. Although we could now end up in the awkward situation where DNS tells us one thing and the (/dns) multiaddr another.

Anyway, at this point I think we don’t really have a problem we’re solving, this seems like premature optimization.

Author: > There’s no round trip cost. ALPN negotiation happens during the TLS handshake.

Yes, you're correct. I was thinking of the Upgrade header which would only happen in `h2c` which would be the rare case (e.g. doing HTTP not on tls).

I'll remove this section.

currently [registered](https://github.com/multiformats/multiaddr), this document

> claims their use.

Reviewer 4: I don't think we should have two sources of truth. Either it is `multiformats/multiaddr` or `libp2p/specs`, but not both. Mind creating a pull request against `multiformats/multiaddr`?

Author: will do after we agree on this.

> If a node wants to advertise specific HTTP versions it may advertise a
> `/http-1.1`, `/http-2`, or `/h3` instead. Although these components are not
> currently [registered](https://github.com/multiformats/multiaddr), this document
> claims their use.

Reviewer: We should not mention things that do not exist, nor encourage people to invent more codes  the rest of the ecosystem is expected to support.

I feel this only muddies the takeaway for the document, could we remove the notion of "you could explicitly state the HTTP version" from this doc? It is not possible with URLs, so maybe replace this with a note:

```suggestion
The end user agent decides on HTTP version to use, based on the multiaddr prefix, application, server negotiation, and specific use case. This follows what existing `http://` URL implementations do.
```

Author: I think it's useful to be able to signal that you support a specific HTTP version so that nodes don't have to waste a round trip in version negotiation. Supporting these is easy, since if you can't or won't act on this information you can interpret these as an opaque HTTP endpoint and fall back to HTTP version negotiation (e.g. in the browser).

If we agree on these I can register these codes in the table.

Author: I'm going to remove this section per discussion in another thread. You don't pay for a round trip from HTTP/1.1 to HTTP/2, and HTTP/3 will be obvious when using a udp port.

## Multiaddr representation

The multiaddr of a node with an HTTP transport ends with `/http` and is prefixed
by information that would let an HTTP client know how to reach the server
(remember that multiaddrs are [interpreted right to
left](https://github.com/multiformats/multiaddr#interpreting-multiaddrs)). 

The following are examples of multiaddrs for HTTP transport capable nodes:

> * `/dns/example.com/tls/http`

Reviewer 4: `/tls/http` is deprecated in favor of `/https` according to https://github.com/multiformats/multiaddr/blob/master/protocols.csv. Is it necessary to list it here?

Author: I went back and forth. I think `/https` is a bit more ergonomic than `/tls/http` (I at least often double check if tls comes before or after). But either way we must still support this since it's in the wild.

Author: I removed `/https`. I think your original comment was a typo since `/https` is deprecated in favor of `/tls/http`.

> * `/dns/example.com/https`

Reviewer 3: Missing IP protocol and port.

Reviewer 3: Is there any precedent for this? Multiaddrs are all about being explicit.

Author: Yes, off the top of my head: https://github.com/multiformats/js-multiaddr-to-uri#usage converts `('/dnsaddr/protocol.ai/https') -> https://protocol.ai`.

This is in line with how to interpret multiaddrs. The stuff left of the `/http` component goes to the HTTP transport. As long as the HTTP transport knows how to handle `/dns/example.com` that's all that's necessary.

Another example is the `webrtc` transport. We don't specify SCTP anywhere in there.

Reviewer 3: That’s because there’s no WebRTC without SCTP, as there also is no QUIC v1 without TLS.

Author: We also don't specify if the webrtc connection is over UDP vs TCP (since I believe you can run it on top of TCP as well). This is handled automatically by the webrtc transport. There's actually a lot we don't specify in the `/webrtc` transport. The only thing that is specified is the bootstrap connection.

But anyways, my earlier point stands. Anything that is valid for an http transport is valid for the components left of `/http`.

Reviewer 3: Actually, I agree. If you're doing DNS, you have SVCB, so that's fine.

* `/ip4/1.2.3.4/tcp/443/https`

> * `/ip4/1.2.3.4/tcp/443/tls/http`

Reviewer 3: Can we define an unambiguous way to encode this? For reference, check what we decided for regarding `/tls/ws` and `/wss`.

Author: They are equivalent. I think what we do in go-libp2p is transform `/wss` to `/tls/ws` for consistency.

Author: `/wss` is derectated just like `/https`. But that doesn't mean it isn't a valid example of a multiaddr you would see in the wild. See also this comment: https://github.com/libp2p/specs/pull/550#discussion_r1213381244.

Author: I could add a note that this is deprecated.

* `/ip6/2001:0db8:85a3:0000:0000:8a2e:0370:7334/tcp/443/tls/http`
* `/ip4/1.2.3.4/udp/50781/tls/http` // We can infer this is an HTTP/3 address

> * `/ip4/1.2.3.4/udp/50781/quic-v1/http` // Not necessary to specify that the

Reviewer 3: Let's not infer things. There might other protocols (and new QUIC versions) that also run on top of UDP.

Author: You're right. We shouldn't include the quic-v1 line. Just the udp and port part and the http client can figure out what HTTP transport to use (which right now only means h3).

Author: What do you mean? This is only for `/http` transports. To deploy a new UDP based transport you would append it to the end of a udp multiaddr (e.g. `/ip4/1.2.3.4/udp/54321/utp)

Author: I believe @marten-seemann's argument is that we should specify this HTTP stack is sitting on quic-v1 (rather than some other future hypothetical UDP based transport).

The general counter argument is that everything left of the `/http` component is handled by the HTTP transport. If the HTTP transport needs more information than the [authority](https://datatracker.ietf.org/doc/html/rfc3986#section-3) and whether we're using tcp vs udp, then the extra information should be supplied. Right now, I don't think there's ambiguity here. If someday there is ambiguity we can add the extra information before the `/http` component. e.g. `/ip4/1.2.3.4/udp/54321/new-transport/http`.

  HTTP endpoint is on top of QUIC, but it is okay if it does.

## HTTP Paths

It may be tempting to add an HTTP path to end of the multiaddr to specify some
information about a user protocol. However the `/http` component is not a user
protocol, and it doesn't accept any parameters. It only signals that a node is
capable of an HTTP transport.

> The HTTP Path exists in the user protocol level. In the HTTP semantics. You can

Reviewer: ```suggestion
The HTTP Path exists in the user protocol level. In the HTTP semantics defined by [RFC 9110](https://httpwg.org/specs/rfc9110.html). You can
```

> The HTTP Path exists in the user protocol level. In the HTTP semantics. You can

Reviewer: ```suggestion
The HTTP Path exists in the user protocol level. HTTP Semantics are transport-agnostic, and defined by [RFC 9110](https://httpwg.org/specs/rfc9110.html). You can
```

> use these semantics on any transport including, but not limited to, the HTTP
> transport.

Reviewer: ```suggestion
use these semantics on any transport including, but not limited to, the HTTP
transports like [HTTP/1.1](https://www.rfc-editor.org/info/rfc7235), [HTTP/2](https://www.rfc-editor.org/info/rfc9113), or [HTTP/3](https://www.rfc-editor.org/info/rfc9114).
```

For example, say you want to fetch a file using the [IPFS trustless

> gateway](https://docs.ipfs.tech/reference/http/gateway/). It may be tempting to

Reviewer: When we have specs, prefer them over any user docs:
```suggestion
gateway](https://specs.ipfs.tech/http-gateways/trustless-gateway/). It may be tempting to
```

> use the following URL to reference a file:

Reviewer 5: I think the message still comes across, but maybe a stronger example is for a multiaddr to a node that speaks the IPFS trustless HTTP gateway API?  It seems more likely that someone advertises "I speak HTTP trustless gateway" rather than "you can fetch a specific file from me".

Given the current landscape, I could imagine someone thinking that `/ip4/127.0.0.1/tcp/8080/http/httppath/ipfs` should map to an HTTP trustless gateway instance.

Author: Thanks, I've reworded this. Resolving this comment, but please reopen if needed.

> use the following URL to reference a file:

Reviewer: ```suggestion
use the following multiaddr to reference a file:
```

transport. That will be defined in a future document.

This exists to clarify the role of the `/http` component in Multiaddrs early to
avoid confusion and conflicting interpretations.

## What is an HTTP transport

An HTTP transport is simply a node that can speak some standardized version of
HTTP. Intuitively if you can `curl` it with HTTP, then it speaks HTTP.

Most environments will have a way to create an HTTP Client and Server, and the
specific HTTP version used will be opaque. We use the `/http` component at the
end of the multidadr to signal that this server supports an HTTP transport. The
end user agent decides on HTTP version to use, based on the multiaddr prefix,
application, server negotiation, and specific use case. This follows what
existing `http://` URL implementations do.

## Multiaddr representation

The multiaddr of a node with an HTTP transport ends with `/http` and is prefixed
by information that would let an HTTP client know how to reach the server
(remember that multiaddrs are [interpreted right to
left](https://github.com/multiformats/multiaddr#interpreting-multiaddrs)). 

The following are examples of multiaddrs for HTTP transport capable nodes:

> * `/dns/example.com/tls/http`

Reviewer 6: If this says `/tls`, does that mean we expect both ends to conform to the libp2p TLS spec with its mandatory extension for signing the public key? Or is this vanilla TLS like a browser would perform with a server?

If this would not end in `/http`, it would definitely mean the former, correct?

But that would make it impossible for existing HTTP clients to connect as they don't understand our mandatory extension.

The same question applies to `/quic-v1`.

I think technically, we could say that - because multiaddresses are interpreted from right to left - the `/http` component defines that `/tls` and `/quic-v1` mean vanilla TLS without our extension. This sounds very confusing to me though.

Author: > If this says /tls, does that mean we expect both ends to conform to the libp2p TLS spec with its mandatory extension for signing the public key? Or is this vanilla TLS like a browser would perform with a server?

The latter, vanilla TLS. Authentication will be optional in the HTTP transport.

> If this would not end in /http, it would definitely mean the former, correct?

I think so because we only support authenticated streams. And we only have defined the libp2p TLS extension as the auth mechanism.

The libp2p TLS extension is about authentication. If we require auth then we need it (unless we define an alternate way to do auth). In the HTTP transport, auth will be optional so we don't require the libp2p TLS extension.

> The same question applies to /quic-v1.

Yup, and hopefully this same answer makes sense.

Reviewer 6: I just realized that the TLS spec only applies to the `/tls/1.0.0` stream protocol and not to the multiaddr transport component.

However, for QUIC, we explicitly define that the TLS authentication is required if the `/quic-v1` component is used: https://github.com/libp2p/specs/tree/master/quic#peer-authentication

I think this confusion would not arise if we would define `/http1`, `/http2` and `/http3`.

Somewhere else you say that we only want to define that a node speaks the HTTP semantics but since we are talking about transports in the multiaddr, wouldn't it make sense to also talk about the various HTTP versions (which are transport related?).

Reviewer 6: > > However, for QUIC, we explicitly define that the TLS authentication is required if the /quic-v1 component is used: [`master`/quic#peer-authentication](https://github.com/libp2p/specs/tree/master/quic?rgh-link-date=2023-06-15T17%3A46%3A48Z#peer-authentication)
> 
> This spec is specific for libp2p streams on QUIC. `/quic-v1` in the context of `/http` can have its own semantics (e.g. optional peer auth).

At least to me, that is completely non-obvious to me from reading the specs.

Multiaddress components currently always mean the same thing, right? This spec would be the first one where we _change_ the meaning of an existing component.

Author: Added in e71d7c5

Most environments will have a way to create an HTTP Client and Server, and the
specific HTTP version used will be opaque. We use the `/http` component at the
end of the multidadr to signal that this server supports an HTTP transport. The
end user agent decides on HTTP version to use, based on the multiaddr prefix,
application, server negotiation, and specific use case. This follows what
existing `http://` URL implementations do.

## Multiaddr representation

The multiaddr of a node with an HTTP transport ends with `/http` and is prefixed
by information that would let an HTTP client know how to reach the server
(remember that multiaddrs are [interpreted right to
left](https://github.com/multiformats/multiaddr#interpreting-multiaddrs)). 

The following are examples of multiaddrs for HTTP transport capable nodes:

* `/dns/example.com/tls/http`
* `/ip4/1.2.3.4/tcp/443/tls/http`
* `/ip6/2001:0db8:85a3:0000:0000:8a2e:0370:7334/tcp/443/tls/http`
* `/ip4/1.2.3.4/udp/50781/quic-v1/http`

Note: When we use `/quic-v1/http` (or any other QUIC version)
implementations MUST use the correct ALPN (e.g. `h3`) and not `libp2p` when

> using the HTTP transport.

Reviewer 3: That applies to any transport, doesn't it?

Author: Correct, I'll mention that as well. Thanks
